src/functions/updateDmsTaskWithDates/framework/dms/table-mapping.ts

```diff
@@ -0,0 +1,170 @@
+/**
+ * Handles generating DMS table mappings from a much simpler logical input.
+ */
+import { getLogger } from '../util';
+import { DateTime } from 'luxon';
+
+const logger = getLogger('table-mapping', 'debug');
+
+export interface OrCondition {
+  operator: string;
+  value?: string;
+  start?: string;
+  end?: string;
+}
+
+export interface AndFilter {
+  column: string;
+  orConditions: OrCondition[];
+}
+
+export interface Table {
+  sourceName: string;
+  andFilters?: AndFilter[];
+  removeColumns?: string[];
+}
+
+export interface Options {
+  sourceSchema: string;
+  destSchema: string;
+  tables: Table[];
+}
+
+export function generateTableMapping(options: Options): any {
+  const config = { rules: [] };
+
+  // first add a transformation rule from source to dest schema name...
+  config.rules.push({
+    'rule-type': 'transformation',
+    'rule-id': '1',
+    'rule-name': '1',
+    'rule-action': 'rename',
+    'rule-target': 'schema',
+    'object-locator': {
+      'schema-name': options.sourceSchema,
+    },
+    value: options.destSchema,
+  });
+
+  let index = 2; // rule 1 is above
+  options.tables.forEach((element: Table) => {
+    // then add include selection rules for every table...
+    const rule: any = {
+      'rule-type': 'selection',
+      'rule-id': `${index}`,
+      'rule-name': `${index}`,
+      'object-locator': {
+        'schema-name': options.sourceSchema,
+        'table-name': element.sourceName,
+      },
+      'rule-action': 'include',
+    };
+
+    if (element.andFilters) {
+      rule['filters'] = [];
+
+      element.andFilters.forEach((andFilter: AndFilter) => {
+        const filter: any = {
+          'filter-type': 'source',
+          'column-name': andFilter.column,
+          'filter-conditions': [],
+        };
+
+        andFilter.orConditions.forEach((condition: OrCondition) => {
```

Comment:
I understand it might be necessary to have this nested looping but can we separate these `forEach` handlers out into their own functions, would make it more readable and testable.

Reply:
refactored